### PR TITLE
Add support for disabling beatmap preview buttons

### PR DESCRIPTION
**Pull request by frenzibyte**
This PR is adding support for disabling play buttons programmatically, which will be used in an upcoming PR.

I'm not entirely sure whether hiding the button completely when disabled is the way, but that's how it's done osu!web-side.

Also in case anyone wonders, the change in `PreviewButton` was because `PlayButton` doesn't do anything in `OnClick` now, `PlayButton.Button` does, and either way that `Click()` method really shouldn't exist, it's way hacky to the point clicks aren't propagated to the drawable's children, if for testing, `InputManager` is more preferable to use.

### CODE REVIEW

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs [hunk 1]
```diff
@@ -74,6 +74,13 @@ private void load(OsuColour colours, OverlayColourProvider colourProvider)
             background.Colour = colourProvider.Background6;
         }
 
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            playButton.Enabled.BindValueChanged(e => Enabled.Value = e.NewValue, true);
```
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs:81]: well, that's the design of `PreviewButton`, it creates a `PlayButton` inside of it, `PlayButton` is the one managing over the beatmap set's preview track, `PreviewButton` only adds to the look of the button. don't think it's feasible to change the direction of propagation without rewriting `PreviewButton` all over.

File: osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs [hunk 2]
```diff
@@ -64,6 +69,17 @@ public PlayButton(BeatmapSetInfo setInfo = null)
...
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            updateEnabledState();
+        }
+
+        private void updateEnabledState()
+        {
+            button.Enabled.Value = BeatmapSet != null;
+        }
```
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs]: well, this isn't actually meant to be modifiable programmatically, only done so in test scene by straight force, so it doesn't matter by much for now.

main PR uses the internal button enabled state this way, if you may ask:
https://github.com/ppy/osu/blob/3f16acaf82cfcd8ad911997bdc1d42eaf046f361/osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs#L84-L99

File: osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs:81]: no, that's not quite the same. the preview track and the playing bindable are re-exposed in `PreviewButton`. it's composition. if it were possible to proxy `Enabled`, i'd recommend that, but that doesn't seem feasible as the preview button already has its own `Enabled`. so it'd have to be declared with `new` which i don't even know if it'll behave correctly since shadowing is confusing.

speaking of which, given this direction of propagation, how do you even disable a `PreviewButton`? setting `Enabled` on the preview doesn't propagate to the play button, and you can't disable the inner button from outside because it's private.

File: osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs]: that is all entirely backwards. all of it. `Enabled` is *public*. it needs to work the same way 99% of all other components work. so it *must* be settable programmatically and it *must* work reliably.

File: osu.Game.Tests/Visual/Online/TestSceneDirectPanel.cs [hunk 3]
```diff
@@ -125,13 +130,38 @@ private void load(RulesetStore rulesets)
...
+            public PlayDisabledGridBeatmapPanel(BeatmapSetInfo beatmap)
+                : base(beatmap)
+            {
+            }
+
+            protected override void LoadComplete()
+            {
+                base.LoadComplete();
+
+                PlayButton.ChildrenOfType<ClickableContainer>().Single().Enabled.Value = false;
```
Comment (bdach) [inline osu.Game.Tests/Visual/Online/TestSceneDirectPanel.cs:163]: this also needs to go away. it should be no more difficult than directly setting `Enabled` on the button to test this.

File: osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs:81]: All good points, does this commit satisfy what you wanted https://github.com/ppy/osu/commit/1734a889b838023ca310bfc3fd5378dc1f22fce2, otherwise not sure I can come up with anything else simple to do, helping would be appreciated.

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs:81]: The direction you have on that commit seems.. better? see comments i left on it.

File: osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs [hunk 4]
```diff
@@ -39,88 +46,50 @@ public BeatmapSetInfo BeatmapSet
...
 
-        protected override void OnHoverLost(HoverLostEvent e)
-        {
-            if (!playing.Value)
-                icon.FadeColour(Color4.White, 120, Easing.InOutQuint);
-            base.OnHoverLost(e);
+            playing.Disabled = disabled;
         }
 
+        public void TogglePlaying() => playing.Toggle();
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs:89]: what's the point? `Playing` (the public one) is... well, public.

File: osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs [hunk 5]
```diff
@@ -74,6 +77,20 @@ private void load(OsuColour colours, OverlayColourProvider colourProvider)
             background.Colour = colourProvider.Background6;
         }
 
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            Playing.BindDisabledChanged(_ => updateEnabledState());
+            Enabled.BindValueChanged(_ => updateEnabledState());
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs:85]: so given that `Enabled` sets `playing.Disabled`, are these both going to fire? this is still way too complicated for how it should be. it shouldn't be this hard to disable a button. i have no clue why that boolean value has to be duplicated in 2 places and moved all over.

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs (hunk 4 above)
Comment (frenzibyte) [reply osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs:89]: that would make `Playing.Disabled` mutable, which is not wanted

File: osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs (hunk 5 above)
Comment (frenzibyte) [reply osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs]: I've just realized `PreviewButton` being a clickable container and exposing an `Enabled` bindable is just useless, so I removed the inheritance and it seems to have improved readability pretty well.

File: osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs [hunk 6]
```diff
@@ -61,9 +61,9 @@ public PreviewButton()
                     Origin = Anchor.Centre,
                     Size = new Vector2(18),
                 },
+                new HoverClickSounds()
```
Comment (bdach) [inline osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs:64]: why go to the trouble of not inheriting `OsuClickableContainer` anymore if you're just going to jam `HoverClickSounds` back in, which is like half of the point of that class?

i swear this is all because of the baffling insistence on storing the button's enabled state in `playing.Disabled` instead of a separate `Enabled` bindable, which is both confusing and just doesn't fit the general design of everything around it (which is apparent by re-implementing button disable logic all over the place). i can't see a reasonable justification for it at all.

Comment (frenzibyte) [reply osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs:64]: no, it's literally because I can't find any way to do this cleanly without extra refactoring, `PlayButton` is disabling itself very innocently on some conditions, and with the current layout, `PreviewButton`, which is containing `PlayButton`, has to be disabled based on whether the play button is.

i've removed inheritance to `OsuClickableContainer` to remove exposing `Enabled` bindable and get rid of hiding properties and `base.Enabled` stuff to improve readability (and even get rid of public `Action` event while there, `OsuClickableContainer` isn't even correct for this type of button).

honestly at this point i'm just thinking about making `PreviewButton` inherit `PlayButton` and expose some method to create extra content around the button, it's the only thing that would make sense and would get this over with.

File: osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs [hunk 7]
```diff
@@ -166,5 +151,89 @@ private void attemptStart()
...
+                return base.OnHover(e);
+            }
+
+            protected override void OnHoverLost(HoverLostEvent e)
+            {
+                updateDisplay();
+                base.OnHoverLost(e);
+            }
+
+            private void updateDisplay()
```
Comment (peppy) [inline osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs:226]: I'd probably make this a `virtual protected` to avoid duplicating all the calling logic of this function (it's being bound in three different events). Make sure to attach xmldoc mentioning it will be called on hover or disabled state changes.

In fact, if you send in hovered and disabled as argument you may be able to keep the disabled bindable private.

File: osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs [hunk 8]
```diff
@@ -39,88 +49,63 @@ public BeatmapSetInfo BeatmapSet
...
 
-        protected override bool OnClick(ClickEvent e)
-        {
-            playing.Toggle();
-            return true;
+            AlwaysDisabled.BindValueChanged(_ => updateEnabledState(), true);
         }
 
-        protected override bool OnHover(HoverEvent e)
+        private void updateEnabledState()
```
Comment (peppy) [inline osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs:88]: Move to end of file with other private methods.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs (hunk 7 above)
Comment (frenzibyte) [reply osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs:226]: Well, one problem is that hovering over the icon button is different from hovering over `HeaderPlayButton`, so the `virtual protected` method would have to accept hovering state of both the `PlayButton` and the icon button, which sounds kinda too much from just seperating them.

**Comment by peppy** [comment]
I've changed some naming to read better, but right now this is a regression in code quality, to the point I can't easily follow what's going on. At very least, the bindable logic between `PlayButton` and its nested class needs to be removed. Just replace it with a method or two non-bindable properties.

**Comment by frenzibyte** [comment]
I agree bindables can make things hard to follow in some cases, will take a look at that.

**Comment by frenzibyte** [comment]
Just to be clear, are you referring to `ShouldDisplay` and `Playing`? as I'm thinking about:
 - a `protected bool` property for each of them.
 - a virtual `UpdateDisplay` method on the main `PlayButton` class called when any of the properties are changed.
 - an `UpdateState` method on the `PlayButton.Button` nested class called when any of the properties are changed.

**Comment by peppy** [comment]
I'm pretty sure you still want the outwards facing ones as bindables? Aren't they being used in such a fashion? My complaint was about the communication between `PlayButton` and `PlayButton.Button`.

**Comment by smoogipoo** [comment]
Going to close due to inactivity + going to require a herculean effort to merge due to the panels having since been rewritten. Best start off fresh.
